### PR TITLE
Feature/rails3 custom fields api

### PR DESCRIPTION
**Pull request by machisuji**
Writing and reading custom fields for work packages (aka planning elements). Reading projects' custom fields. Querying existing custom fields.

Issues:
- https://www.openproject.org/issues/2564
- https://www.openproject.org/issues/2586

### CODE REVIEW

**Comment by mfrister** [comment]
Just fixed the changelog conflict, let's see what Travis thinks of this now (after one week of laying around).

**Comment by ulferts** [comment]
I'd like to have the performance implications assessed before we merge this. @apfelfabrik is working hard to improve performance for the timelines. If this branch would endanger that goal, e.g. by introducing n+1 queries, I would not like to have that merged.

**Comment by machisuji** [comment]
What's still missing is eager loading of custom fields in /api/v2/projects/index. Will do it later today.

**Comment by machisuji** [comment]
Now using eager loading for custom fields (i.e. projects, types and translations) in:
- projects/index
- projects/planning_element_custom_fields
- custom_fields/index

File: spec/controllers/api/v2/planning_elements_controller_spec.rb [hunk 1]
```diff
@@ -367,29 +367,67 @@ def self.become_member_with_delete_planning_element_permissions(&block)
...
+          :project_id => project.identifier,
+          :format => 'xml',
+          :planning_element => planning_element.attributes.merge(:custom_fields => [
+            { :id => custom_field.id, :value => "Wurst" }])
+        response.response_code.should == 303
+
+        id = response.headers["Location"].scan(/\d+/).last.to_i
+        get 'show', :project_id => project.identifier, :id => id, :format => 'json'
+
+        response.response_code.should == 200
```
Comment (hschink) [inline spec/controllers/api/v2/planning_elements_controller_spec.rb]: `response.should be_success` is a shorter form.

File: spec/controllers/api/v2/planning_elements_controller_spec.rb [hunk 2]
```diff
@@ -480,25 +518,71 @@ def expect_redirect_to
...
+      end
+
+      let(:planning_element) do
+        FactoryGirl.create(
+          :work_package,
+          :author => author,
+          :type => type,
+          :project => project,
+          :custom_values => [
+            CustomValue.new(:custom_field => custom_field, :value => "Mett")])
```
Comment (hschink) [inline spec/controllers/api/v2/planning_elements_controller_spec.rb]: Use a factory to create objects your test depends on.

File: spec/controllers/api/v2/planning_elements_controller_spec.rb [hunk 3]
```diff
@@ -367,29 +367,67 @@ def self.become_member_with_delete_planning_element_permissions(&block)
...
+      it 'creates a new planning element with the given custom field value' do
+        post 'create',
+          :project_id => project.identifier,
+          :format => 'xml',
+          :planning_element => planning_element.attributes.merge(:custom_fields => [
+            { :id => custom_field.id, :value => "Wurst" }])
+        response.response_code.should == 303
+
+        id = response.headers["Location"].scan(/\d+/).last.to_i
+        get 'show', :project_id => project.identifier, :id => id, :format => 'json'
```
Comment (hschink) [inline spec/controllers/api/v2/planning_elements_controller_spec.rb]: You are testing two tings here:
1. custom field creation
2. custom field fetching

Thus, you should write (1) a test that creates a custom field via a post request and check that the field exists in the database and (2) a test that fetches custom fields that you created with the help of a factory.

There are other occasions where you use this pattern too.

**Comment by hschink** [comment]
Looks good! Just adapt the tests and I will merge this PR.

File: doc/CHANGELOG.md [hunk 4]
```diff
@@ -29,6 +29,7 @@ See doc/COPYRIGHT.rdoc for more details.
 
 # Changelog
 
+* `#2564` Support custom fields in REST API v2 for work packages and projects
```
Comment (hschink) [inline doc/CHANGELOG.md:32]: Ticket [2564](https://www.openproject.org/issues/2586) is missing

**Pull request merged**